plugin/pulpcore/plugin/stages/profiler.py

```diff
@@ -0,0 +1,169 @@
+from asyncio import Queue
+import pathlib
+import time
+import uuid
+
+from rq.job import get_current_job
+
+from pulpcore.tasking import connection
+
+
+CONN = None
+
+
+class ProfilingQueue(Queue):
+    """
+    A customized subclass of asyncio.Queue that records time in the queue and between queues.
+
+    This Profiler records some data on items that are inserted and removed from Queues. This data is
+    stored on items in a dictionary attribute called 'extra_data'. If this attribute does not exist
+    on an item, the ProfileQueue adds it.
+
+    The following statistics are computed for each Queue and the stage that it feeds into:
+
+        * waiting time - The number of seconds an item waited in the Queue for this stage.
+        * service time - The number of seconds an item received service in this stage.
+        * queue_length - The number of waiting items in the queue, measured before each new arrival.
+        * interarrival_time - The number of seconds since the previous arrival to this Queue.
+
+    See the :meth:`create_profile_db_and_connection()` docs for more info on the database tables and
+    layout.
+
+    Args:
+         stage_uuid (uuid.UUID): The uuid of the stage this ProfilingQueue delivers work into.
+         args (tuple): unused positional arguments
+         kwargs (dict): unused keyword arguments
+    """
+
+    def __init__(self, stage_uuid, *args, **kwargs):
+        self.last_arrival_time = time.time()
+        self.stage_uuid = stage_uuid
+        return super().__init__(*args, **kwargs)
+
+    def get_nowait(self):
+        """
+        Thinly wrap `asyncio.get_nowait` and record when get_nowait() operations happen.
+        """
+        item = super().get_nowait()
+        if item:
+            now = time.time()
+            if not hasattr(item, 'extra_data'):
+                # track stages that use QuerySet items too
+                item.extra_data = {}
+            item.extra_data['last_waiting_time'] = now - item.extra_data['last_put_time']
```

Comment:
I think L#52 never executes and L#50-L#53 can be removed. Items are always sent to put() first because you have to put an item in the queue first before you can get() it. Even if get() is called first in actuality due to concurrency run-order, L#47 call to super() will not return until an item is available at which point we can be guaranteed a put() has occurred.

This was helpful via review. Ty. I'm removing these lines.